Loop/src/loop_device.hxx

```diff
@@ -261,6 +264,108 @@ public:
     } // for rank
   }
 
+  // Loop over the outermost "boundary" points in the interior. They correspond
+  // to points that are shifted inwards by = cctk_nghostzones[3] from those that
+  // CarpetX identifies as boundary points. From the perspective of CarpetX (or
+  // AMReX), these do not belong in the outer boundary, but rather the interior.
+  // This excludes ghost faces, but includes ghost edges/corners on non-ghost
+  // faces. Loop over faces first, then edges, then corners. Modified from
+  // loop_bnd_device.
+  template <int CI, int CJ, int CK, int VS = 1, int N = 1,
+            int NT = AMREX_GPU_MAX_THREADS, typename F>
+  inline CCTK_ATTRIBUTE_ALWAYS_INLINE void
+  loop_int_bnd_device(const vect<int, dim> &group_nghostzones,
+                      const F &f) const {
+    // boundary_box sets bnd_min and bnd_max
+    vect<int, dim> bnd_min, bnd_max;
+    // if on (Carpetx) boundary points, then
+    //   bnd_min = nghostzones - lbnd
+    //   bnd_max = gsh - offset - nghostzones - lbnd
+    //   where lbnd : an array of cctk dim integers containing the lowest index
+    //   (in each direction) of the local grid, as seen on the global grid
+    // else,
+    //   bnd_min = -100000000
+    //   bnd_max = gsh + 1000000
+    // where cctk_gsh : an array of cctk dim integers with the global grid size
+    boundary_box<CI, CJ, CK>(group_nghostzones, bnd_min, bnd_max);
+
+    // domain_boxes sets all_min, all_max, int_min and int_max
+    vect<int, dim> all_min, all_max, int_min, int_max;
+    // Indices for the entire grid
+    //   all_min = ghost_offset;
+    //   all_max = lsh - offset - ghost_offset;
+    // Indices for the interior
+    //   int_min = nghostzones;
+    //   int_max = lsh - offset - nghostzones;
+    domain_boxes<CI, CJ, CK>(group_nghostzones, all_min, all_max, int_min,
+                             int_max);
+
+    // Shift the indices towards the interior by nghostzones
+    all_min += nghostzones;
```

Comment:
The variable `nghostzones` contains the default number of ghost zones. Individual grid functions can override this, and the actual number of ghost zones for the loop is passed in `group_nghostzones`. In most cases both with be the same.

I don't think you're handling the case correct when these values are different. One way to "fix" this would be to check whether they are the same, and to abort if they are not.

Reply:
I just added ```assert(group_nghostzones[d] == nghostzones[d]``` to check for this.